File: dkb/format_dkb.go
package dkb

import (
	"errors"
	"rahul-rakshit/formatbankcsv/output"
	"rahul-rakshit/formatbankcsv/utils"
)

func formatDkb(inputLines [][]string) ([][]string, error) {
	expectedInputHeader := []string{"Buchungsdatum", "Wertstellung", "Status", "Zahlungspflichtige*r", "Zahlungsempfänger*in", "Verwendungszweck", "Umsatztyp", "Betrag", "Gläubiger-ID", "Mandatsreferenz", "Kundenreferenz"}
	outputLines := [][]string{
		output.Header,
	}

	for index, inputLine := range inputLines {
		if index == 0 {
			if !utils.IsLineEqual(inputLine, expectedInputHeader) {
				return [][]string{}, errors.New("Unexpected header for dkb format")
			}

			continue
		}

		date, _ := convertDkbDate(inputLine[0])
		reference := inputLine[5]
		umsatztyp := inputLine[6]
		amount, _ := convertDkbAmount(inputLine[7])
		var vendor string

		if umsatztyp == "Ausgang" {
			vendor = inputLine[4]
		} else {
			vendor = inputLine[3]
		}

		outputLines = append(outputLines, []string{date, vendor, reference, amount})
	}

	return utils.SortByDate(outputLines), nil
}

File: dkb/formatDkb.ts
import { Header } from "../utils/constants.ts";
import { isLineEqual } from "../utils/isLineEqual.ts";
import { sortByDate } from "../utils/sortByDate.ts";
import { convertDkbAmount } from "./convertDkbAmount.ts";
import { convertDkbDate } from "./convertDkbDate.ts";

const expectedInputHeader = [
  "Buchungsdatum",
  "Wertstellung",
  "Status",
  "Zahlungspflichtige*r",
  "Zahlungsempfänger*in",
  "Verwendungszweck",
  "Umsatztyp",
  "IBAN",
  "Betrag (€)",
  "Gläubiger-ID",
  "Mandatsreferenz",
  "Kundenreferenz",
];

export function formatDkb(inputLines: string[][]): string[][] {
  const outputLines: string[][] = [[...Header]];

  for (let i = 0; i < inputLines.length; i++) {
    const inputLine = inputLines[i];

    if (i === 0) {
      if (!isLineEqual(inputLine, expectedInputHeader)) {
        throw new Error("Unexpected header for dkb format");
      }
      continue;
    }

    const date = convertDkbDate(inputLine[0]);
    const reference = inputLine[5];
    const umsatztyp = inputLine[6];
    const amount = convertDkbAmount(inputLine[8]);
    const vendor = umsatztyp === "Ausgang" ? inputLine[4] : inputLine[3];

    outputLines.push([date, vendor, reference, amount]);
  }

  return sortByDate(outputLines);
}
